src/communication/v2/session.hpp

```diff
@@ -0,0 +1,331 @@
+      return false;
+    }
+    boost::asio::dispatch(this->strand_, [shared_this = this->shared_from_this(), data, len, have_more] {
+      shared_this->DoWrite(data, len, have_more);
+    });
+    return true;
+  }
+
+  bool IsConnected() const {
+    return std::visit(
+        utils::Overloaded{
+            [this](const SSLSocket &socket) { return execution_active_ && socket.lowest_layer().is_open(); },
+            [this](const TCPSocket &socket) { return execution_active_ && socket.is_open(); }},
+        socket_);
+  }
+
+ private:
+  explicit Session(tcp::socket &&socket, TSessionData *data, ServerContext &server_context, tcp::endpoint endpoint,
+                   int inactivity_timeout_sec, std::string_view service_name)
+      : socket_(CreateWebSocket(std::move(socket), server_context)),
+        strand_{boost::asio::make_strand(GetExecutor())},
+        output_stream_([this](const uint8_t *data, size_t len, bool have_more) { return Write(data, len, have_more); }),
+        session_(data, endpoint, input_buffer_.read_end(), &output_stream_),
+        endpoint_{endpoint},
+        remote_endpoint_{GetRemoteEndpoint()},
+        service_name_{service_name},
+        timeout_seconds_(inactivity_timeout_sec),
+        timeout_timer_(GetExecutor()) {
+    std::visit(
+        utils::Overloaded{[](SSLSocket &socket) {
+                            socket.lowest_layer().set_option(boost::asio::ip::tcp::no_delay(true));  // enable PSH
+                            socket.lowest_layer().set_option(
+                                boost::asio::socket_base::keep_alive(true));  // enable SO_KEEPALIVE
+                            socket.lowest_layer().non_blocking(false);
+                          },
+                          [](TCPSocket &socket) {
+                            socket.set_option(boost::asio::ip::tcp::no_delay(true));        // enable PSH
+                            socket.set_option(boost::asio::socket_base::keep_alive(true));  // enable SO_KEEPALIVE
+                            socket.non_blocking(false);
+                          }},
+        socket_);
+    timeout_timer_.expires_at(boost::asio::steady_timer::time_point::max());
+    spdlog::info("Accepted a connection from {}:", service_name_, remote_endpoint_.address(), remote_endpoint_.port());
+  }
+
+  void DoWrite(const uint8_t *data, size_t len, bool /*have_more*/) {
+    if (!IsConnected()) {
+      return;
+    }
+    std::visit(utils::Overloaded{[shared_this = this->shared_from_this(), data, len](TCPSocket &socket) {
+                                   boost::asio::write(socket, boost::asio::buffer(data, len));
+                                 },
+                                 [shared_this = this->shared_from_this(), data, len](SSLSocket &socket) {
+                                   boost::asio::write(socket, boost::asio::buffer(data, len));
+                                 }},
+               socket_);
+  }
+
+  void DoRead() {
+    if (!IsConnected()) {
```

Comment:
Do we want this to be a `no-op` if there is no active connection?

Reply:
What do you mean by that?

Reply:
I mean that `DoRead` doesn't really tell you if it read anything because if the connection is false it returns immediately. The question is, are these the right semantics? (I am not sure that's why I am asking)

Reply:
Aha, I get the question, so we don't want to do anything if for some reason we initiated a shutdown, but maybe there will be read/write as the next action, we want to skip it since we will be closing the connection soon.